src/inmanta/server/services/fileservice.py

```diff
@@ -42,131 +42,86 @@ def __init__(self) -> None:
-                )
+        return file.content
 
     @handle(methods.stat_files)
     async def stat_files(self, files: List[str]) -> Apireturn:
         """
         Return which files in the list exist on the server
         """
-        return 200, {"files": self.stat_file_internal(files)}
+        return 200, {"files": await self.stat_file_internal(files)}
 
-    def stat_file_internal(self, files: Iterable[str]) -> List[str]:
+    async def stat_file_internal(self, files: Iterable[str]) -> List[str]:
         """
         Return which files in the list don't exist on the server
         """
-        # A dict is used here instead of a set to have efficient set-like behaviour while preserving insertion order. Only its
-        # keys are relevant.
-        response: dict[str, object] = {}
-
-        for f in files:
-            f_path = os.path.join(self.server_slice._server_storage["files"], f)
-            if not os.path.exists(f_path):
-                response[f] = None
-
-        return list(response.keys())
+        return list(await File.get_non_existing_files(files))
 
     @handle(methods.diff)
     async def file_diff(self, a: str, b: str) -> Apireturn:
         """
         Diff the two files identified with the two hashes
         """
-        if a == "" or a == "0":
-            a_lines: List[str] = []
-        else:
-            a_path = os.path.join(self.server_slice._server_storage["files"], a)
-            if not os.path.exists(a_path):
-                raise NotFound()
-
-            with open(a_path, "r", encoding="utf-8") as fd:
-                a_lines = fd.readlines()
-
-        if b == "" or b == "0":
-            b_lines: List[str] = []
-        else:
-            b_path = os.path.join(self.server_slice._server_storage["files"], b)
-            if not os.path.exists(b_path):
-                raise NotFound()
-
-            with open(b_path, "r", encoding="utf-8") as fd:
-                b_lines = fd.readlines()
 
+        def _parse_like_readlines(content: str) -> list[str]:
+            """
+            Parse the given string like <file_handle>.readlines() would do.
+            """
+            lines = content.split("\n")
+            if lines[-1] == "":
+                # file ended with a newline character
+                lines = lines[0:-1]
+            # Add newline characters that were removed by spit()
+            return [f"{r}\n" for r in lines]
+
+        async def _get_lines_for_file(content_hash: str) -> list[str]:
+            if content_hash == "" or content_hash == "0":
+                return []
+            else:
+                file = await File.get_one(content_hash=content_hash)
+                if not file:
+                    raise NotFound()
+                file_content = file.content.decode(encoding="utf-8")
+                return _parse_like_readlines(file_content)
```

Comment:
```suggestion
                # keepends for backwards compatibility with <file_handle>.readlines()
                return file_content.splitlines(keepends=True)
```
I believe this is equivalent, and less brittle in case of CR line breaks (they should be rare but I'm not sure if we can rule them out).

Or is there a difference I'm not seeing?

Reply:
Oh. I forgot this comment.

Reply:
Done